// react-redux-provide/lib/keyConcats.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.unshiftClear = exports.pushClear = exports.unshiftWait = exports.pushWait = exports.unshiftReplicator = exports.pushReplicator = exports.unshiftReplication = exports.pushReplication = exports.unshiftOnReady = exports.pushOnReady = exports.unshiftOnInstantiated = exports.pushOnInstantiated = exports.unshiftEnhancer = exports.pushEnhancer = exports.unshiftMiddleware = exports.pushMiddleware = undefined;

var _createKeyConcat = require('./createKeyConcat');

var _createKeyConcat2 = _interopRequireDefault(_createKeyConcat);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var middlewareKey = 'middleware';
var pushMiddleware = exports.pushMiddleware = (0, _createKeyConcat2.default)(middlewareKey);
var unshiftMiddleware = exports.unshiftMiddleware = (0, _createKeyConcat2.default)(middlewareKey, true);

var enhancerKey = 'enhancer';
var pushEnhancer = exports.pushEnhancer = (0, _createKeyConcat2.default)(enhancerKey);
var unshiftEnhancer = exports.unshiftEnhancer = (0, _createKeyConcat2.default)(enhancerKey, true);

var onInstantiatedKey = 'onInstantiated';
var pushOnInstantiated = exports.pushOnInstantiated = (0, _createKeyConcat2.default)(onInstantiatedKey);
var unshiftOnInstantiated = exports.unshiftOnInstantiated = (0, _createKeyConcat2.default)(onInstantiatedKey, true);

var onReadyKey = 'onReady';
var pushOnReady = exports.pushOnReady = (0, _createKeyConcat2.default)(onReadyKey);
var unshiftOnReady = exports.unshiftOnReady = (0, _createKeyConcat2.default)(onReadyKey, true);

var replicationKey = 'replication';
var pushReplication = exports.pushReplication = (0, _createKeyConcat2.default)(replicationKey);
var unshiftReplication = exports.unshiftReplication = (0, _createKeyConcat2.default)(replicationKey, true);

var replicatorKey = [replicationKey, 'replicator'];
var pushReplicator = exports.pushReplicator = (0, _createKeyConcat2.default)(replicatorKey);
var unshiftReplicator = exports.unshiftReplicator = (0, _createKeyConcat2.default)(replicatorKey, true);

var waitKey = 'wait';
var pushWait = exports.pushWait = (0, _createKeyConcat2.default)(waitKey);
var unshiftWait = exports.unshiftWait = (0, _createKeyConcat2.default)(waitKey, true);

var clearKey = 'clear';
var pushClear = exports.pushClear = (0, _createKeyConcat2.default)(clearKey);
var unshiftClear = exports.unshiftClear = (0, _createKeyConcat2.default)(clearKey, true);
